parser.c3

```diff
@@ -0,0 +1,658 @@
+    runtime_assert(cursor:try_skip_next_and_surrounding_white_space("return"));
+    return {cursor:parse_expression()};
+}
+
+function parse_line : (cursor : Cursor&) -> Line = {
+    // TODO: this should be templated?
+    cursor:try_skip_white_space();
+    if cursor:is_next("if") {
+        let pointer : If& = cursor.allocator:allocate<If>();
+        pointer = cursor:parse_if();
+        return {3, ref_to_addr(&pointer)};
+    }
+    if cursor:is_next("while") {
+        let pointer : While& = cursor.allocator:allocate<While>();
+        pointer = cursor:parse_while();
+        return {4, ref_to_addr(&pointer)};
+    }
+    if cursor:is_next("for") {
+        let pointer : For& = cursor.allocator:allocate<For>();
+        pointer = cursor:parse_for();
+        return {5, ref_to_addr(&pointer)};
+    }
+    if cursor:is_next("return") {
+        let pointer : Return& = cursor.allocator:allocate<Return>();
+        pointer = cursor:parse_return();
+        return {2, ref_to_addr(&pointer)};
+    }
+    if cursor:is_next("{") {
+        let pointer : Scope& = cursor.allocator:allocate<Scope>();
+        pointer = cursor:parse_scope();
+        return {6, ref_to_addr(&pointer)};
+    }
+
+    // The line is not a statement, it must be an expression or a assignment
+    // TODO:
+    runtime_assert(false);
+}
+
+function parse_scope : (cursor : Cursor&) -> Scope = {
+    runtime_assert(cursor:try_skip_next_and_surrounding_white_space("{"));
+
+    let result : Scope = {make_vector<Line>()};
+    while !cursor:try_skip_next_and_surrounding_white_space("}") {
+        if cursor:get_char() == "\0"[0] {
+            syntax_error(current_char.position, "TODO: unterminated '}'");
+        }
+        result.lines:push_back(cursor:parse_line());
+    }
+    return result;
+}
+
+function parse_graphene_function : (cursor : Cursor&) -> FunctionDefinition = {
+    runtime_assert(cursor:try_skip_next_and_surrounding_white_space("function"));
+
+    // Generic definitions
+    let generics : Vector<GenericDefinition>; // Uninitialized
+
+    if cursor:has_next("[") {
+        generics = cursor:parse_generic_definition_list();
+    } else {
+        // Empty the vector if the function has no generic arguments
+        generics = make_vector<GenericDefinition>(cursor.allocator);
+    }
+
+    cursor:try_skip_white_space();
+
+    // The function's name
+    let name : ProgramSlice = cursor:get_next_ascii_only_identifier({comma: false, operators: false, colon: true});
+    if !cursor:try_skip_next_and_surrounding_white_space(":") {
+        syntax_error(cursor.position, "TODO: expected a ':'");
+    }
+
+    let args : Vector<Argument> = cursor:parse_comma_separated_named_members(make_span(")"));
+
+    if !cursor:try_skip_next_and_surrounding_white_space("->") {
+        syntax_error(cursor.position, "TODO: expected '->'");
+    }
+
+    let return_type : Type = cursor:parse_type();
+    let scope : Scope = cursor:parse_scope();
```

Comment:
Did you drop the `=`?

Reply:
Not intentionally :stuck_out_tongue_winking_eye: